forem/forem
# Add mention autocomplete to article comments

## What type of PR is this? (check all applicable)


## Description

This PR adds the user mention autocomplete functionality to comments (and replies to comments) on posts, including from the notifications view.

## Related Tickets & Documents

RFC: [Create mention autocomplete](https://github.com/forem/rfcs/pull/16/files)

## QA Instructions, Screenshots, Recordings

We need to test this on both mobile and desktop devices. Both comments and replies to comments should be tested, as they are initialised differently. 

If working correctly, when you enter the comment box and start typing `@..` plus some letters it should:

- Start showing you results (or a message saying no matches) after a minimum of 2 characters after the `@`
- Update your search results dynamically as you type or delete characters
- Never show more than 6 options in the dropdown (you might need to add some users with similar usernames to your DB to be able to test this)
- Return results based purely on username, but show you the user's name and avatar in the dropdown too
- Show the dropdown below where your cursor started typing the mention (desktop)
- Show the dropdown at the start of the line below where you started typing the mention (mobile)
- Reposition the dropdown if the text area wrapped to a new line while you were typing the search (desktop and mobile)
- Complete the text `@username` when you select one of the options by tapping/clicking
- Scroll through the suggestions using the up and down arrow keys, and complete the `@username` text when you select an option by pressing enter
- Close the dropdown temporarily if you press Escape
- Exit the search/dropdown fully if you type a space 
- Exit the search/dropdown fully if you click elsewhere on the page
- See below for additional accessibility considerations

Recording of me testing out some of this on desktop:


https://user-images.githubusercontent.com/20773163/110127694-d4f44000-7dbd-11eb-992f-f49b49d0f662.mp4

And on mobile-size screen:


https://user-images.githubusercontent.com/20773163/110127727-e0e00200-7dbd-11eb-8b67-2c386a50777d.mp4



### UI accessibility concerns?

Definite accessibility implications with this change. As the user types `@` the focus leaves the normal textarea and moves to a 'combobox' (which informs an assistive tech user that they can now search/select as well as just type). 

We need to make sure this is working when using keyboard alone, including the ability to simply 'keep typing' and ignore the suggestions completely. In addition we should check with a screen reader that:

- When the user starts typing `@...` the screen reader announces they are now in a `combobox` and how to interact with it
- The user can navigate through the listed options with the keyboard as above, and they are announced appropriately
- Selected option is announced when user presses Enter

I've recorded myself navigating through using this on Safari with VoiceOver (output shown in text):


https://user-images.githubusercontent.com/20773163/110127500-a24a4780-7dbd-11eb-96cb-1c53e8487af9.mp4



## Added tests?

- [X] Yes: Cypress tests added

## [Forem core team only] How will this change be communicated?

      or in a [forem.dev](http://forem.dev) post


## [optional] Are there any post deployment tasks we need to perform?

## [optional] What gif best describes this PR or how it makes you feel?

![Lionel Richie signs hello](https://i.giphy.com/media/WS3i2y88foYpE584rI/giphy.webp)
